com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -104,29 +116,160 @@ public int AddEntry(ulong networkObjectId, int behaviourIndex, int variableIndex
+            writer.WriteUInt16(entry.Key.BehaviourIndex);
+            writer.WriteUInt16(entry.Key.VariableIndex);
+            writer.WriteUInt16(entry.Key.TickWritten);
+            writer.WriteUInt16(entry.Position);
+            writer.WriteUInt16(entry.Length);
+        }
+
+        /// <summary>
+        /// Read a received Entry
+        /// Must match WriteEntry
+        /// </summary>
+        /// <param name="reader">The readed to read the entry from</param>
+        internal Entry ReadEntry(NetworkReader reader)
+        {
+            Entry entry;
+            entry.Key.NetworkObjectId = reader.ReadUInt64();
+            entry.Key.BehaviourIndex = reader.ReadUInt16();
+            entry.Key.VariableIndex = reader.ReadUInt16();
+            entry.Key.TickWritten = reader.ReadUInt16();
+            entry.Position = reader.ReadUInt16();
+            entry.Length = reader.ReadUInt16();
+            entry.Fresh = false;
+
+            return entry;
+        }
+
         /// <summary>
         /// Allocate memory from the buffer for the Entry and update it to point to the right location
         /// </summary>
         /// <param name="entry">The entry to allocate for</param>
         /// <param name="size">The need size in bytes</param>
-        public void AllocateEntry(ref Entry entry, long size)
+        public void AllocateEntry(ref Entry entry, int index, int size)
         {
             // todo --M1--
             // this will change once we start reusing the snapshot buffer memory
             // todo: deal with free space
             // todo: deal with full buffer
 
-            entry.Position = (ushort)FreeMemoryPosition;
+            int pos;
+
+            if (entry.Length > 0)
+            {
+                Allocator.Deallocate(index);
+            }
+
+            bool ret = Allocator.Allocate(index, size, out pos);
+
+            if (!ret)
+            {
+                //todo: error handling
+            }
+
+            entry.Position = (ushort)pos;
             entry.Length = (ushort)size;
-            FreeMemoryPosition += (int)size;
+        }
+
+        /// <summary>
+        /// Read the buffer part of a snapshot
+        /// Must match WriteBuffer
+        /// The stream is actually a memory stream and we seek to each variable position as we deserialize them
+        /// </summary>
+        /// <param name="reader">The NetworkReader to read our buffer of variables from</param>
+        /// <param name="snapshotStream">The stream to read our buffer of variables from</param>
+        internal void ReadBuffer(NetworkReader reader, Stream snapshotStream)
+        {
+            int snapshotSize = reader.ReadUInt16();
+
+            snapshotStream.Read(Buffer, 0, snapshotSize);
+
+            for (var i = 0; i < LastEntry; i++)
+            {
+                if (Entries[i].Fresh && Entries[i].Key.TickWritten > 0)
+                {
+                    var nv = FindNetworkVar(Entries[i].Key);
```

Comment:
This will fail if a single NetworkObject has changed between the client and the server.

Example.
1. Client sends variable for NetworkObject with Id 5 (PlayerPrefab)
2. Server Despawns NetworkObjectId with Id 5
3. Server spawns (NetworkOtherPrefab), Id 5 gets assigned to that new object.
4. This function tries to read the snapshot, searches for a variable which does not exist on "NetworkOtherPrefab"

Reply:
Hmm, I was prepared to write a very similar comment, until this popped out.

Reply:
Ok, good point, seems like I made assumptions about uniqueness of Ids which don't hold.

Reply:
How did that not manifest in existing mlapi ? It seems the whole serialization makes the same assumptions, no ?

Reply:
Right now this is not handled in MLAPI. There is an id recycling delay which in theory should prevent the above situation in many cases. But that's just one of the cases. Another problem would be if the object just does not exist anymore due to being despawned/destroyed.

I don't think this PR should/can address all those issues. Just wanted to point them out.

Reply:
Noted.